firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java

```diff
@@ -191,17 +191,51 @@ static SnapshotVersion getPostMutationVersion(MutableDocument document) {
    * @return A map of fields to transform results.
    */
   protected Map<FieldPath, Value> localTransformResults(
-      Timestamp localWriteTime, MutableDocument mutableDocument) {
+      Timestamp localWriteTime, MutableDocument mutableDocument, Mutation mutation) {
     Map<FieldPath, Value> transformResults = new HashMap<>(fieldTransforms.size());
     for (FieldTransform fieldTransform : fieldTransforms) {
       TransformOperation transform = fieldTransform.getOperation();
       Value previousValue = mutableDocument.getField(fieldTransform.getFieldPath());
+      if (mutation != null) {
```

Comment:
I wonder if we need this if we pass ObjectValue to this method instead of MutableDocument+mutation. Would this work?

```
diff --git a/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java b/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java
index f5ea818e..934868a6 100644
--- a/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java
+++ b/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Mutation.java
@@ -191,19 +191,11 @@ public abstract class Mutation {
    * @return A map of fields to transform results.
    */
   protected Map<FieldPath, Value> localTransformResults(
-      Timestamp localWriteTime, MutableDocument mutableDocument, Mutation mutation) {
+      Timestamp localWriteTime, ObjectValue currentValue) {
     Map<FieldPath, Value> transformResults = new HashMap<>(fieldTransforms.size());
     for (FieldTransform fieldTransform : fieldTransforms) {
       TransformOperation transform = fieldTransform.getOperation();
-      Value previousValue = mutableDocument.getField(fieldTransform.getFieldPath());
-      if (mutation != null) {
-        FieldUpdate update = getFieldUpdate(fieldTransform.getFieldPath());
-        if (update.type == FieldUpdate.Type.DELETE) {
-          previousValue = null;
-        } else if (update.type == FieldUpdate.Type.SET) {
-          previousValue = update.value;
-        }
-      }
+      Value previousValue = currentValue.get(fieldTransform.getFieldPath());
       transformResults.put(
           fieldTransform.getFieldPath(), transform.applyToLocalView(previousValue, localWriteTime));
     }
diff --git a/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/PatchMutation.java b/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/PatchMutation.java
index ed93bb97..06eb8fad 100644
--- a/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/PatchMutation.java
+++ b/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/PatchMutation.java
@@ -14,6 +14,8 @@
 
 package com.google.firebase.firestore.model.mutation;
 
+import androidx.annotation.Nullable;
+
 import com.google.firebase.Timestamp;
 import com.google.firebase.firestore.model.DocumentKey;
 import com.google.firebase.firestore.model.FieldPath;
@@ -134,10 +136,12 @@ public final class PatchMutation extends Mutation {
       return;
     }
 
-    Map<FieldPath, Value> transformResults = localTransformResults(localWriteTime, document, null);
     ObjectValue value = document.getData();
     value.setAll(getPatch());
+
+    Map<FieldPath, Value> transformResults = localTransformResults(localWriteTime, value);
     value.setAll(transformResults);
+
     document
         .convertToFoundDocument(getPostMutationVersion(document), document.getData())
         .setHasLocalMutations();
@@ -145,7 +149,7 @@ public final class PatchMutation extends Mutation {
 
   @Override
   public Mutation squash(
-      Mutation baseMutation, MutableDocument document, Timestamp localWriteTime) {
+          @Nullable Mutation baseMutation, MutableDocument document, Timestamp localWriteTime) {
     MutableDocument preconditionTest =
         (baseMutation != null && (baseMutation instanceof DeleteMutation))
             ? MutableDocument.newNoDocument(getKey(), document.getVersion())
@@ -156,7 +160,7 @@ public final class PatchMutation extends Mutation {
     }
 
     Map<FieldPath, Value> transformResults =
-        localTransformResults(localWriteTime, document, baseMutation);
+        localTransformResults(localWriteTime, baseMutation.getValue());
 
     ObjectValue mergedPatch =
         (baseMutation != null && baseMutation.getValue() != null)
```

Reply:
I think this only works when `baseMutation` is `Set`. When it is `Patch`, the transforms need to be applied on top of fields from `document` + `baseMutation`.

Passing `getValue` from `baseMutation` will missing the field values we need from `document`, and the transforms will be applied as if those fields are missing if `baseMutation.getValue()` contains no such fields.

Reply:
I am trying to get this to work. I think I am closeish but I still have more test failures than before. Will play around with it more tomorrow. See https://github.com/firebase/firebase-android-sdk/pull/2966 for a broken draft.